### PR TITLE
feat(calls): Impl call recording

### LINKED ISSUE #1178: feat(calling): add option to start/stop recording a call
using these functions from Blink

image

### PR DESCRIPTION
**Pull request by Flemmli97**
### What this PR does 📖

- Adds ability to record the current call via button in the current call window

### Which issue(s) this PR fixes 🔨

- Resolve #1178

### CODE REVIEW

**Review by sdwoodbury — APPROVED** [review]

**Comment by phillsatellite** [comment]
When I tested on all OS's the option to record a call goes away once the User you are calling answers. Is that intended or should the option to record a call be there the entire call? 

https://github.com/Satellite-im/Uplink/assets/93608357/6dadd107-8195-40fc-a9cd-31c9ef65828a

**Comment by Flemmli97** [comment]
oops. ye no. fixed now

File: ui/src/components/media/calling.rs [hunk 1]
```diff
@@ -178,6 +182,48 @@ fn ActiveCallControl(cx: Scope<ActiveCallProps>) -> Element {
...
+                                .to_string(),
+                            rsp: tx,
+                        })) {
+                            log::error!("failed to send blink command: {e}");
+                            continue;
+                        }
+
+                        match rx.await {
+                            Ok(_) => {
+                                *recording.write() = true;
```
Comment (lgmarchi) [inline ui/src/components/media/calling.rs]: Not a blocker, but it is recommended to use `with_mut` on hooks to change status

File: ui/src/components/media/calling.rs [hunk 2]
```diff
@@ -178,6 +182,48 @@ fn ActiveCallControl(cx: Scope<ActiveCallProps>) -> Element {
...
+                        if let Err(e) =
+                            warp_cmd_tx.send(WarpCmd::Blink(BlinkCmd::StopRecording { rsp: tx }))
+                        {
+                            log::error!("failed to send blink command: {e}");
+                            continue;
+                        }
+
+                        match rx.await {
+                            Ok(_) => {
+                                *recording.write() = false;
```
Comment (lgmarchi) [inline ui/src/components/media/calling.rs]: Same here

**Review by lgmarchi — APPROVED** [review]

**Comment by phillsatellite** [comment]
@Flemmli97 recording button is available during entire call now, thanks! 💥  I wanted to ask a question though, after you are done recording call, should we be saving the recording file somewhere? I wanted to try to listen back to it after I was done recording but couldnt seem to find the file anywhere. Wasnt sure if we plan on adding that in a future PR 

https://github.com/Satellite-im/Uplink/assets/93608357/fa42d6a7-d8a2-46a7-b31f-0ca2d2f7c1de

**Comment by dariusc93** [comment]
> @Flemmli97 recording button is available during entire call now, thanks! 💥 I wanted to ask a question though, after you are done recording call, should we be saving the recording file somewhere? I wanted to try to listen back to it after I was done recording but couldnt seem to find the file anywhere. Wasnt sure if we plan on adding that in a future PR
> Screen.Recording.2023-09-20.at.11.24.20.AM.mov

Based on the code, it looks like it would be saving in a `recordings` directory. Likely in the uplink directory unless a path is defined to a different location (?).

**Comment by Flemmli97** [comment]
> > @Flemmli97 recording button is available during entire call now, thanks! 💥 I wanted to ask a question though, after you are done recording call, should we be saving the recording file somewhere? I wanted to try to listen back to it after I was done recording but couldnt seem to find the file anywhere. Wasnt sure if we plan on adding that in a future PR
> > Screen.Recording.2023-09-20.at.11.24.20.AM.mov
> 
> Based on the code, it looks like it would be saving in a `recordings` directory. Likely in the uplink directory unless a path is defined to a different location (?).

yes they are saved in the uplink/recordings dir. i did notice though that for me the file does not work but that may be on blink?

**Comment by phillsatellite** [comment]
Hey @Flemmli97 so I just tested again and I found where the recordings are saved and they actually play back! Very cool! But one issue I noticed, when the user hits the record button there doesnt seem to be any indication that user is recording. Other than that UI issue this is starting to work very well. Awesome feature!! 🔥 

And something Sara had mentioned that I think is a good idea, maybe in a future PR we can add a place in the UI where the user can more easily find the saved recordings 

https://github.com/Satellite-im/Uplink/assets/93608357/76eb109e-addd-45fd-90f4-47467871895f

**Review by dariusc93 — APPROVED** [review]

**Review by phillsatellite — APPROVED** [review]

**Pull request merged**